pandas/core/algorithms.py

```diff
@@ -104,92 +246,41 @@ def match(to_match, values, na_sentinel=-1):
+    table = htable(min(len(to_match), 1000000))
+    table.map_locations(values)
+    result = table.lookup(to_match)
 
     if na_sentinel != -1:
 
         # replace but return a numpy array
         # use a Series because it handles dtype conversions properly
-        from pandas.core.series import Series
+        from pandas import Series
         result = Series(result.ravel()).replace(-1, na_sentinel).values.\
             reshape(result.shape)
 
     return result
 
 
-def _match_object(values, index, table_type, type_caster):
-    values = type_caster(values)
-    index = type_caster(index)
-    table = table_type(min(len(index), 1000000))
-    table.map_locations(index)
-    return table.lookup(values)
-
-
-def unique(values):
-    """
-    Compute unique values (not necessarily sorted) efficiently from input array
-    of values
-
-    Parameters
-    ----------
-    values : array-like
-
-    Returns
-    -------
-    uniques
-    """
-    values = com._asarray_tuplesafe(values)
-
-    f = lambda htype, caster: _unique_object(values, htype, caster)
-    return _hashtable_algo(f, values)
-
-
-def _unique_object(values, table_type, type_caster):
-    values = type_caster(values)
-    table = table_type(min(len(values), 1000000))
-    uniques = table.unique(values)
-    return type_caster(uniques)
-
-
 def unique1d(values):
     """
     Hash table-based unique
     """
-    if np.issubdtype(values.dtype, np.floating):
-        table = htable.Float64HashTable(len(values))
-        uniques = np.array(table.unique(_ensure_float64(values)),
-                           dtype=np.float64)
-    elif np.issubdtype(values.dtype, np.datetime64):
-        table = htable.Int64HashTable(len(values))
-        uniques = table.unique(_ensure_int64(values))
-        uniques = uniques.view('M8[ns]')
-    elif np.issubdtype(values.dtype, np.timedelta64):
-        table = htable.Int64HashTable(len(values))
-        uniques = table.unique(_ensure_int64(values))
-        uniques = uniques.view('m8[ns]')
-    elif np.issubdtype(values.dtype, np.signedinteger):
-        table = htable.Int64HashTable(len(values))
-        uniques = table.unique(_ensure_int64(values))
-    elif np.issubdtype(values.dtype, np.unsignedinteger):
-        table = htable.UInt64HashTable(len(values))
-        uniques = table.unique(_ensure_uint64(values))
-    else:
-
-        # its cheaper to use a String Hash Table than Object
-        if lib.infer_dtype(values) in ['string']:
-            table = htable.StringHashTable(len(values))
-        else:
-            table = htable.PyObjectHashTable(len(values))
+    values = _ensure_arraylike(values)
+    original = values
+    htable, _, values, dtype, ndtype = _get_hashtable_algo(values)
 
-        uniques = table.unique(_ensure_object(values))
+    table = htable(len(values))
+    uniques = table.unique(values)
+    uniques = _reconstruct_data(uniques, dtype, original)
 
     return uniques
 
 
+unique = unique1d
```

Comment:
This is exposed at the top-level, so can you keep the docstring of `unique` ? (put it in `unique1D`)